### PR TITLE
bugfix & optional tensorflow implementation

### PR DESCRIPTION
**Pull request by valeriosofi**
- Support ignore_compilers also for torchscript and tflite
- Bug on quantization half onnx: posix path instead of string
- Bug on tensor rt quantization: numpy array passed to inference learner instead of tensor.
- Tensorflow now will not be installed by default by nebullvm

### CODE REVIEW

File: nebullvm/api/functions.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
 def _get_dl_framework(model: Any):
     if isinstance(model, torch.nn.Module):
         return DeepLearningFramework.PYTORCH
-    elif isinstance(model, tf.Module):
+    elif isinstance(model, tf.Module) and tf.Module != object:
```
Comment (morgoth95) [inline nebullvm/api/functions.py:40]: if you set `tf.Module` to `None` when tensorflow is not installed, you can avoid checking tf.Module type and just check if `model is not None`. Otherwise, we can simply check `model!=object` instead of `tf.Module != object`.

File: nebullvm/compressors/intel.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
 def _get_model_framework(model: Any) -> str:
     if isinstance(model, torch.nn.Module):
         return "torch"
-    elif isinstance(model, tf.Module):
+    elif isinstance(model, tf.Module) and tf.Module != object:
```
Comment (morgoth95) [inline nebullvm/compressors/intel.py:25]: see comment above

File: nebullvm/converters/converters.py [hunk 3]
```diff
@@ -77,13 +77,13 @@ def convert(
                 input_data=input_data,
             )
             return save_path / onnx_name
-        elif isinstance(model, tf.Module):
+        elif isinstance(model, tf.Module) and tf.Module != object:
```
Comment (morgoth95) [inline nebullvm/converters/converters.py:80]: see comment above

File: nebullvm/converters/converters.py [hunk 4]
```diff
@@ -77,13 +77,13 @@ def convert(
...
             convert_tf_to_onnx(
                 model=model,
                 output_file_path=save_path / onnx_name,
             )
             return save_path / onnx_name
-        elif isinstance(model, tf.keras.Model):
+        elif isinstance(model, tf.keras.Model) and tf.keras.Model != object:
```
Comment (morgoth95) [inline nebullvm/converters/converters.py:86]: see comment above

File: nebullvm/converters/converters.py [hunk 5]
```diff
@@ -119,7 +119,7 @@ def convert(
                 input_data=input_data,
             )
             return [model, str(onnx_path)]
-        elif isinstance(model, tf.Module):
+        elif isinstance(model, tf.Module) and tf.Module != object:
```
Comment (morgoth95) [inline nebullvm/converters/converters.py:122]: see comment above

File: nebullvm/pipelines/steps.py [hunk 6]
```diff
@@ -613,7 +613,7 @@ def _get_optimizer_step(
 ) -> Step:
     if isinstance(model, torch.nn.Module):
         return TorchOptimizerStep(logger=logger)
-    elif isinstance(model, tf.Module):
+    elif isinstance(model, tf.Module) and tf.Module != object:
```
Comment (morgoth95) [inline nebullvm/pipelines/steps.py:616]: see comment above

File: nebullvm/pipelines/steps.py [hunk 7]
```diff
@@ -622,7 +622,7 @@ def _get_optimizer_step(
 def _get_pipeline_name(model: Any):
     if isinstance(model, torch.nn.Module):
         return "pytorch_pipeline"
-    elif isinstance(model, tf.Module):
+    elif isinstance(model, tf.Module) and tf.Module != object:
```
Comment (morgoth95) [inline nebullvm/pipelines/steps.py:625]: see comment above

File: nebullvm/transformations/precision_tfms.py [hunk 8]
```diff
@@ -33,7 +33,7 @@ def _transform(self, _input: Any, **kwargs) -> Any:
                 if _input.dtype == torch.float32
                 else _input
             )
-        elif isinstance(_input, tf.Tensor):
+        elif isinstance(_input, tf.Tensor) and tf.Tensor != object:
```
Comment (morgoth95) [inline nebullvm/transformations/precision_tfms.py:36]: see comment above

File: nebullvm/utils/benchmark.py [hunk 9]
```diff
@@ -4,21 +4,23 @@
...
+from nebullvm.utils.optional_modules import tensorflow as tf
 from nebullvm.utils.tf import create_model_inputs_tf
 from nebullvm.utils.torch import create_model_inputs_torch
 
 
 def _get_dl_framework(model: Any):
     if isinstance(model, torch.nn.Module) or str(model).startswith("Pytorch"):
         return DeepLearningFramework.PYTORCH
-    elif isinstance(model, tf.Module) or str(model).startswith("Tensorflow"):
+    elif (isinstance(model, tf.Module) and tf.Module != object) or str(
```
Comment (morgoth95) [inline nebullvm/utils/benchmark.py:21]: see comment above

File: nebullvm/utils/onnx.py [hunk 10]
```diff
@@ -2,16 +2,16 @@
...
 
 from nebullvm.base import InputInfo, DataType, DeepLearningFramework
+from nebullvm.utils.optional_modules import tensorflow as tf
 
 
 def convert_to_numpy(tensor: Any):
     if isinstance(tensor, torch.Tensor):
         tensor = tensor.cpu().detach().numpy()
-    elif isinstance(tensor, tf.Tensor):
+    elif isinstance(tensor, tf.Tensor) and tf.Tensor != object:
```
Comment (morgoth95) [inline nebullvm/utils/onnx.py:14]: see comment above

File: nebullvm/utils/tf.py [hunk 11]
```diff
@@ -1,15 +1,14 @@
...
 from nebullvm.base import InputInfo, DataType
+from nebullvm.utils.optional_modules import tensorflow as tf
 
 
 def get_outputs_sizes_tf(
     tf_model: Union[tf.Module, tf.keras.Model], input_tensors: List[tf.Tensor]
 ) -> List[Tuple[int, ...]]:
     outputs = tf_model(*input_tensors)
-    if isinstance(outputs, tf.Tensor):
+    if isinstance(outputs, tf.Tensor) and tf.Tensor != object:
```
Comment (morgoth95) [inline nebullvm/utils/tf.py:11]: see comment above

File: nebullvm/utils/tf.py [hunk 12]
```diff
@@ -34,6 +33,6 @@ def run_tf_model(
     model: tf.Module, input_tensors: Tuple[tf.Tensor]
 ) -> Tuple[tf.Tensor]:
     pred = model.predict(*input_tensors)
-    if isinstance(pred, tf.Module):
+    if isinstance(pred, tf.Module) and tf.Module != object:
```
Comment (morgoth95) [inline nebullvm/utils/tf.py:36]: see comment above

File: nebullvm/api/functions.py [hunk 13]
```diff
@@ -199,6 +199,8 @@ def optimize_model(
     FEEDBACK_COLLECTOR.start_collection(model, framework=dl_framework)
     if metric_drop_ths is not None and metric_drop_ths <= 0:
         metric_drop_ths = None
+    elif metric_drop_ths is not None:
```
Comment (morgoth95) [inline nebullvm/api/functions.py:202]: shouldn't we check if `metric` is None as well?

File: nebullvm/pipelines/steps.py [hunk 14]
```diff
@@ -374,11 +374,11 @@ class TorchOptimizerStep(OptimizerStep):
     def _get_optimizers(
         self, ignore_compilers: List[ModelCompiler]
     ) -> Dict[ModelCompiler, BaseOptimizer]:
-        optimizers = {
-            ModelCompiler.TORCHSCRIPT: PytorchBackendOptimizer(
+        optimizers = dict()
```
Comment (morgoth95) [inline nebullvm/pipelines/steps.py:377]: use `{}` instead of `dict()`

File: nebullvm/pipelines/steps.py [hunk 15]
```diff
@@ -462,11 +462,11 @@ class TFOptimizerStep(OptimizerStep):
     def _get_optimizers(
         self, ignore_compilers: List[ModelCompiler]
     ) -> Dict[ModelCompiler, BaseOptimizer]:
-        optimizers = {
-            ModelCompiler.TFLITE: TensorflowBackendOptimizer(
+        optimizers = dict()
```
Comment (morgoth95) [inline nebullvm/pipelines/steps.py:465]: see comment above

File: nebullvm/pipelines/steps.py [hunk 16]
```diff
@@ -659,7 +657,7 @@ def build_pipeline_from_model(
         logger (Logger, optional): Logger defined by the user.
     """
     compressor_step = _get_compressor_step(
```
Comment (morgoth95) [inline nebullvm/pipelines/steps.py:659]: why don't we pass the metric information anymore?

File: nebullvm/api/functions.py (hunk 13 above)
Comment (valeriosofi) [reply nebullvm/api/functions.py:202]: Yep you are right

File: nebullvm/pipelines/steps.py (hunk 16 above)
Comment (valeriosofi) [reply nebullvm/pipelines/steps.py:659]: Because it was used only on the check that I removed that prevented the compressor to be used when the metric was none. Now if metric_drop_ths is set the metric is always != None, so it's not necessary anymore
